sys/event/event.c

```diff
@@ -14,24 +28,38 @@
 #include "clist.h"
 #include "thread.h"
 
+void event_queue_init_detached(event_queue_t *queue)
+{
+    assert(queue);
+    memset(queue, '\0', sizeof(*queue));
+}
+
 void event_queue_init(event_queue_t *queue)
 {
     assert(queue);
     memset(queue, '\0', sizeof(*queue));
     queue->waiter = (thread_t *)sched_active_thread;
 }
 
+void event_queue_claim(event_queue_t *queue)
+{
+    assert(queue && (queue->waiter == NULL));
+    queue->waiter = (thread_t *)sched_active_thread;
+}
+
 void event_post(event_queue_t *queue, event_t *event)
 {
-    assert(queue && queue->waiter && event);
+    assert(queue && event);
 
     unsigned state = irq_disable();
     if (!event->list_node.next) {
         clist_rpush(&queue->event_list, &event->list_node);
     }
     irq_restore(state);
 
-    thread_flags_set(queue->waiter, THREAD_FLAG_EVENT);
+    if (queue->waiter) {
```

Comment:
Hm, there is a race now.
Please save ```queue->waiter``` in a temp variable before the irq_restore, the use that variable in the if and thread_flags_set() call.

Reply:
Hmm, not sure that will solve it, e.g.:
- we read 'temp := NULL` inside the critical section
- then we get scheduled away right after irq_restore
- the then active thread calls `event_queue_claim()` and `even_wait()`
- we get scheduled again
- as `temp` is NULL, `thread_flags_set()` is not called, hence the thread waiting on the queue is not notified.

My previous thought was to simply move the complete if statement into the critical section. If I read the code correctly, thread_flags_set() should also work with interrupts disabled, simply postponing the context switch to whenever they are enabled again, right? Am I missing something?

Reply:
Wouldn't the outcome be the same if the thread_flags_set call is inside the critical section? If at the time the critical section starts queue->waiter is NULL, no thread gets woken up. As is, the function will always wake up the thread that was set (if any) at the time the critical section starts.

I was pointing at the crash that is definitely going to happen when queue->waiter changes to NULL before used as an argument to thread_flags_set().

> thread_flags_set() should also work with interrupts disabled

It *should*, but it would also call "thread_yield_higher()" with IRQ's dieabled. No problemo on ARM, AFAIR there was a bug in native messing things up. Also, AVR8 used to hard switch (with no regards to irq state).

Reply:
> I was pointing at the crash that is definitely going to happen when queue->waiter changes to NULL before used as an argument to thread_flags_set().

I see. seems we had different understandings of the API (my fault, should have better documented): it is currently not possible, that the `waiter` will be changed to `NULL`, i.e. a queue can only be claimed once and is then after owned (there is no API function to de-claim a queue). So the problem described above does not really exist, right?

Reply:
> So the problem described above does not really exist, right?

True. Do you mind if we use a temporray variable anyways? I think it is what the compiler would do (without using volatile) anyways. I can provide a patch.

Reply:
Now I am confused: the problem I pointed out above is not solved by a temp variable, and it seems to me that moving the 'if-block' inside the crit section would be the most viable solution. But if that does crash on native (and possibly other platforms), is there anything else we could do?

Reply:
The temporary variable avoids a possible null pointer deref. (waiter is non-null within critical section, but becomes null after irq_restore but before dereferencing it).

The problem you pointed out (event_post() enters critical section with queue unclaimed leads to not waking up a thread) exists both when thread_flags_set() is within or after the critical section, if a temporary variable is used.

If "thread_flag_set()" is not within the critical section, it might happen that at the time event_post() is called, Thread A has the queue claimed, and that changes just before the thread_flags_set() call to another thread (or NULL). In this case, the wrong thread might be flagged. Moving "thread_flags_set()" into the critical section would mitigate that problem, but that causes other problems.

Thus let's fix the null ptr deref -> add the temporary variable. Once we confirmed ```thread_yield_higher()``` is safe to call within critical sections on all platforms, we revisit.

Reply:
> The temporary variable avoids a possible null pointer deref. (waiter is non-null within critical section, but becomes null after irq_restore but before dereferencing it

But this problem is non-existent as we talked about above (waiter can not be changed to NULL once set). So IMHO we do not need to do anything about it.

> Thus let's fix the null ptr deref -> add the temporary variable. Once we confirmed thread_yield_higher() is safe to call within critical sections on all platforms, we revisit.

Though dangerous, as we tend to forget about these re-visits, I agree with this, or else this PR will be stalled forever... So it is down to putting in a TODO note in the code and opening two adressing issues (1 for `thread_yield_higher() is safe to call in crit section` and 1 for adapting this code once it is verfied). Do you agree?

Reply:
> But this problem is non-existent as we talked about above (waiter can not be changed to NULL once set). So IMHO we do not need to do anything about it.

Please add the temp variable. That possible null-pointer deref turns on the alarms of my brain's concurrency checker whenever I look at it. Also, core code should be exemplary. ;) I don't care about the note.

Reply:
not arguing against the temp var (which I will add), just wanted to make sure we have the same understanding about that particular error case...

And as we want to be exemplary here, I think we actually should care quite a bit about the special case of loosing a wakeup call! But anyway, I will open those issues and add a note to make sure its clearly marked.